lib/HLSL/DxilPrecisePropagatePass.cpp

```diff
@@ -51,73 +54,169 @@ class DxilPrecisePropagatePass : public ModulePass {
 private:
   void PropagatePreciseOnFunctionUser(
       Function &F, DxilTypeSystem &typeSys,
-      std::unordered_set<Instruction *> &processedSet);
+      ValueSet &processedSet);
 };
 
 char DxilPrecisePropagatePass::ID = 0;
 
 }
 
 static void PropagatePreciseAttribute(Instruction *I, DxilTypeSystem &typeSys,
-    std::unordered_set<Instruction *> &processedSet);
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeOnPointer(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeOnPointerUsers(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ValueSet &processedSet);
+static void PropagatePreciseAttributeThroughGEPs(
+    Value *Ptr, DxilTypeSystem &typeSys, LLVMContext &Context,
+    ArrayRef<Value*> idxList, ValueSet &processedGEPs,
+    ValueSet &processedSet);
 
 static void PropagatePreciseAttributeOnOperand(
     Value *V, DxilTypeSystem &typeSys, LLVMContext &Context,
-    std::unordered_set<Instruction *> &processedSet) {
-  Instruction *I = dyn_cast<Instruction>(V);
-  // Skip none inst.
-  if (!I)
+    ValueSet &processedSet) {
+  // Skip values already marked.
+  if (!processedSet.insert(V).second)
     return;
 
-  FPMathOperator *FPMath = dyn_cast<FPMathOperator>(I);
-  // Skip none FPMath
-  if (!FPMath)
-    return;
+  if (V->getType()->isPointerTy()) {
+    PropagatePreciseAttributeOnPointer(V, typeSys, Context, processedSet);
+  }
 
-  // Skip inst already marked.
-  if (processedSet.count(I) > 0)
+  Instruction *I = dyn_cast<Instruction>(V);
+  if (!I)
     return;
-  // TODO: skip precise on integer type, sample instruction...
-  processedSet.insert(I);
+
   // Set precise fast math on those instructions that support it.
   if (DxilModule::PreservesFastMathFlags(I))
     DxilModule::SetPreciseFastMathFlags(I);
 
   // Fast math not work on call, use metadata.
-  if (CallInst *CI = dyn_cast<CallInst>(I))
-    HLModule::MarkPreciseAttributeWithMetadata(CI);
+  if (isa<FPMathOperator>(I) && isa<CallInst>(I))
+    HLModule::MarkPreciseAttributeWithMetadata(cast<CallInst>(I));
   PropagatePreciseAttribute(I, typeSys, processedSet);
 }
 
+static void PropagatePreciseAttributeThroughGEPs(
```

Comment:
There is a lot of gep handling code here. I think it could use a high-level comment describing what is going on and what is special about gep that it requires this extra handling.

Reply:
I've written up a high-level description, and also fixed an additional case I thought of (root or intermediate GEP pointer used in function).  This still doesn't propagate through bitcasts or address space casts, so propagation through groupshared array (for instance) will still not occur.  And propagation through UAV load/store would be even further out of the question.